traffic_ops/traffic_ops_golang/login/login.go

```diff
@@ -115,6 +115,47 @@ func LoginHandler(db *sqlx.DB, cfg config.Config) http.HandlerFunc {
 	}
 }
 
+func TokenLoginHandler(db *sqlx.DB, cfg config.Config) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		defer r.Body.Close()
+		var t tc.UserToken
+		if err := json.NewDecoder(r.Body).Decode(&t); err != nil {
+			api.HandleErr(w, r, nil, http.StatusBadRequest, fmt.Errorf("Invalid request: %v", err), nil)
+			return
+		}
+
+		tokenMatches, username, err := auth.CheckLocalUserToken(t.Token, db, time.Duration(cfg.DBQueryTimeoutSeconds)*time.Second)
+		if err != nil {
+			sysErr := fmt.Errorf("Checking token: %v", err)
+			errCode := http.StatusInternalServerError
+			api.HandleErr(w, r, nil, errCode, nil, sysErr)
+			return
+		} else if !tokenMatches {
+			userErr := errors.New("Invalid token. Please contact your administrator.")
+			errCode := http.StatusUnauthorized
+			api.HandleErr(w, r, nil, errCode, userErr, nil)
+			return
+		}
+
+		expiry := time.Now().Add(time.Hour * 6)
+		cookie := tocookie.New(username, expiry, cfg.Secrets[0])
+		httpCookie := http.Cookie{Name: "mojolicious", Value: cookie, Path: "/", Expires: expiry, HttpOnly: true}
+		http.SetCookie(w, &httpCookie)
+		respBts, err := json.Marshal(tc.CreateAlerts(tc.SuccessLevel, "Successfully logged in."))
+		if err != nil {
+			sysErr := fmt.Errorf("Marshaling response: %v", err)
+			errCode := http.StatusInternalServerError
+			api.HandleErr(w, r, nil, errCode, nil, sysErr)
+			return
+		}
+
+		w.Header().Set(tc.ContentType, tc.ApplicationJson)
+		w.Write(append(respBts, '\n'))
```

Comment:
?

Reply:
You asking about the appended `'\n'` ? That's for POSIX compliance. A lot of clients tend to treat the API as a service for obtaining JSON files - hence the historic `.json` suffix on endpoints. But without a trailing newline, they aren't files.

Reply:
makes sense but its not a file download? its a json response ? Do we do this across our other json responses?

Reply:
No, but I don't think it hurts to add it - without a guarantee, of course. Either way it's valid JSON, and the logic for appending a newline is a single call to a builtin with a rune constant, so it shouldn't be hard to maintain.

Reply:
+1 on `\n`.

The trailing newline is still valid JSON, and makes it also a valid POSIX file. @ocket8888 has convinced me we ought to add it to all our endpoints, there's just no reason not to.